docker/Dockerfile.build

```diff
@@ -1,56 +1,101 @@
-# Builder image
-FROM rust:buster as builder
-
-RUN apt-get update && \
-	apt-get install -y --no-install-recommends \
-	make cmake g++ gcc llvm libclang-dev clang ca-certificates
-
-RUN mkdir /zebra
-WORKDIR /zebra
-
-ARG SHORT_SHA
-ENV SHORT_SHA $SHORT_SHA
+# This steps implement cargo-chef for docker layer caching
+# We are using four stages:
+# - chef: installs cargo-chef
+# - planner: computes the recipe file
+# - builder: caches our dependencies and builds the binary
+# - tester: builds and run tests
+# - runtime: is our runtime environment
+FROM rust:bullseye as chef
+RUN cargo install cargo-chef --locked
+WORKDIR /app
+
+FROM chef AS planner
+COPY . .
+RUN cargo chef prepare --recipe-path recipe.json
+
+FROM chef AS builder
+COPY --from=planner /app/recipe.json recipe.json
+
+# Install zebra build deps
+RUN apt-get -qq update && \
+    apt-get -qq install -y --no-install-recommends \
+    llvm \
+    libclang-dev \
+    clang \
+    ca-certificates \
+    ; \
+    rm -rf /var/lib/apt/lists/* /tmp/*
+
+# Install google OS Config agent
+RUN apt-get -qq update && \
+    apt-get -qq install -y --no-install-recommends \
+    curl \
+    lsb-release \
+    ; \
+    echo "deb http://packages.cloud.google.com/apt google-compute-engine-$(lsb_release -cs)-stable main" > /etc/apt/sources.list.d/google-compute-engine.list && \
+    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add - && \
+    apt-get -qq update  && \
+    apt-get -qq install -y --no-install-recommends google-osconfig-agent && \
+    rm -rf /var/lib/apt/lists/* /tmp/*
+
+# Optimize builds. In particular, regenerate-stateful-test-disks.yml was reaching the
+# GitHub Actions time limit (6 hours), so we needed to make it faster.
+ENV RUSTFLAGS -O
+ENV CARGO_HOME /app/.cargo/
+# Build dependencies - this is the caching Docker layer!
+RUN cargo chef cook --release --features enable-sentry --recipe-path recipe.json
+
+ARG RUST_BACKTRACE=1
+ENV RUST_BACKTRACE ${RUST_BACKTRACE}
 
-ENV RUST_BACKTRACE full
-ENV CARGO_HOME /zebra/.cargo/
+COPY . .
+# Pre-download Zcash Sprout and Sapling parameters
+RUN cargo run --locked --release --features enable-sentry --bin zebrad download
```

Comment:
Will this be cached somehow? I don't fully understand how caching works in Docker & GH Actions, but if the source tree changes won't it invalidate the cache in the `COPY` above and thus it will download the parameters every single time? (Sorry, not sure if that's happening, and what the solution would be)

Reply:
It can be invalidated by some specific changes, see https://github.com/ZcashFoundation/zebra/issues/3448

If you think it's better to address that in a separate PR, then your idea seems good for now. (To be clear, the end goal is to make some change so that the parameters are _not_ downloaded on each build, otherwise we'll trigger the server's throttling and we will get CI errors).

Reply:
I'd like to create a reusable workflow for this specific action. I can do it here, or create a separate PR after this has been merged.

I'll leave the decision to @teor2345 and @dconnolly, if they'd like to have it implemented before merging or creating a PR to have it solved in the following 1-3 days.

Reply:
Notes from the engineering meeting:

If we merge this PR without caching, some builds will fail due to parameter download rate-limits.

Reply:
As is, this will run us into the server throttling less frequently than we are now, so I'd be good to merge and quickly follow up with a smaller PR to avoid cache invalidation even more

Reply:
Solved with: https://github.com/ZcashFoundation/zebra/pull/3431/commits/acc38fb6b00a908d6549e887f4797c82aa5af940